--
-- PostgreSQL database dump
--

-- Dumped from database version 13.2
-- Dumped by pg_dump version 13.2

CREATE SCHEMA IF NOT EXISTS variantstore;

SET SCHEMA 'variantstore';

--
-- Name: project; Type: TABLE; Schema: variantstore;
--

CREATE TABLE project (
                         id bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
                         identifier character varying NOT NULL
);

INSERT INTO project(identifier) VALUES ('PROJ') RETURNING id;

--
-- Name: annotationsoftware; Type: TABLE; Schema: variantstore;
--

CREATE TABLE annotationsoftware (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    name character varying,
    version character varying,
    doi character varying
);

INSERT INTO annotationsoftware(name, version, doi) VALUES ('snpeff','4.3t','10.4161/fly.19695');

--
-- Name: annotationsoftware_consequence; Type: TABLE; Schema: variantstore;
--

CREATE TABLE annotationsoftware_consequence (
    annotationsoftware_id bigint NOT NULL,
    consequence_id bigint NOT NULL
);

INSERT INTO annotationsoftware_consequence(annotationsoftware_id, consequence_id)
VALUES (1, 1),
       (1, 2),
       (1, 3),
       (1, 4),
       (1, 5),
       (1, 6),
       (1, 7),
       (1, 8),
       (1, 9),
       (1, 10),
       (1, 11),
       (1, 12),
       (1, 13),
       (1, 14),
       (1, 15),
       (1, 16),
       (1, 17),
       (1, 18),
       (1, 19),
       (1, 20),
       (1, 21),
       (1, 22),
       (1, 23),
       (1, 24),
       (1, 25),
       (1, 26),
       (1, 27),
       (1, 28),
       (1, 29),
       (1, 30),
       (1, 31),
       (1, 32),
       (1, 33),
       (1, 34),
       (1, 35),
       (1, 36),
       (1, 37),
       (1, 38),
       (1, 39),
       (1, 40),
       (1, 41),
       (1, 42),
       (1, 43),
       (1, 44),
       (1, 45),
       (1, 46),
       (1, 47),
       (1, 48),
       (1, 49),
       (1, 50),
       (1, 51),
       (1, 52),
       (1, 53),
       (1, 54),
       (1, 55),
       (1, 56),
       (1, 57),
       (1, 58),
       (1, 59),
       (1, 60),
       (1, 61),
       (1, 62),
       (1, 63),
       (1, 64),
       (1, 65),
       (1, 66),
       (1, 67),
       (1, 68),
       (1, 69),
       (1, 71),
       (1, 72),
       (1, 73),
       (1, 74),
       (1, 75),
       (1, 76),
       (1, 77),
       (1, 78),
       (1, 79),
       (1, 80),
       (1, 81),
       (1, 82),
       (1, 83),
       (1, 84),
       (1, 85),
       (1, 86),
       (1, 87),
       (1, 88),
       (1, 89),
       (1, 90),
       (1, 91),
       (1, 92),
       (1, 93),
       (1, 94),
       (1, 95);

--
-- Name: consequence; Type: TABLE; Schema: variantstore;
--

CREATE TABLE consequence (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    allele character varying,
    codingchange character varying,
    transcriptid character varying,
    transcriptversion integer,
    type character varying,
    biotype character varying,
    canonical boolean,
    aachange character varying,
    cdnaposition character varying,
    cdsposition character varying,
    proteinposition character varying,
    proteinlength integer,
    cdnalength integer,
    cdslength integer,
    impact character varying,
    exon character varying,
    intron character varying,
    strand integer,
    genesymbol character varying,
    featuretype character varying,
    distance integer,
    warnings character varying
);

INSERT INTO consequence(allele, codingchange, transcriptid, transcriptversion, type, biotype, canonical, aachange,
                        cdnaposition, cdsposition, proteinposition, proteinlength, cdnalength, cdslength, impact, exon,
                        intron, strand, genesymbol, featuretype, distance, warnings) VALUES
                        ('G','c.742A>C','ENST00000334528',-1,'missense_variant','protein_coding',false,'p.Asn248His','742','742','248',1196,12355,3591,'MODERATE','1/17','',0,'FMN1','transcript',-1,''),
                        ('G','c.742A>C','ENST00000558197',-1,'missense_variant','protein_coding',false,'p.Asn248His','1020','742','248',503,4060,1512,'MODERATE','1/2','',0,'FMN1','transcript',-1,''),
                        ('G','c.2044-2069A>C','ENST00000559047',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'FMN1','region_of_interest:Mediates_interaction_with_alpha-catenin',-1,''),
                        ('G','c.2044-2069A>C','ENST00000559047',-1,'intron_variant','protein_coding',false,'','','','',-1,-1,-1,'MODIFIER','2/17','',0,'FMN1','transcript',-1,''),
                        ('G','c.1868-59068A>C','ENST00000561249',-1,'intron_variant','protein_coding',false,'','','','',-1,-1,-1,'MODIFIER','1/15','',0,'FMN1','transcript',-1,''),
                        ('G','n.239+651A>C','ENST00000559150',-1,'intron_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','1/1','',0,'FMN1','transcript',-1,''),
                        ('T','c.403G>A','ENST00000552197',-1,'missense_variant','protein_coding',false,'p.Glu135Lys','1088','403','135',503,2599,1512,'MODERATE','7/17','',0,'SLC38A1','transcript',-1,''),
                        ('T','c.403G>A','ENST00000398637',-1,'missense_variant','protein_coding',false,'p.Glu135Lys','1098','403','135',487,8066,1464,'MODERATE','7/17','',0,'SLC38A1','transcript',-1,''),
                        ('T','c.403G>A','ENST00000549049',-1,'missense_variant','protein_coding',false,'p.Glu135Lys','966','403','135',487,3196,1464,'MODERATE','6/16','',0,'SLC38A1','transcript',-1,''),
                        ('T','c.403G>A','ENST00000439706',-1,'missense_variant','protein_coding',false,'p.Glu135Lys','1206','403','135',487,3436,1464,'MODERATE','8/18','',0,'SLC38A1','transcript',-1,''),
                        ('T','c.403G>A','ENST00000546893',-1,'missense_variant','protein_coding',false,'p.Glu135Lys','866','403','135',487,2332,1464,'MODERATE','7/17','',0,'SLC38A1','transcript',-1,''),
                        ('T','c.403G>A','ENST00000546893',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'SLC38A1','topological_domain:Cytoplasmic',-1,''),
                        ('T','n.460G>A','ENST00000549633',-1,'non_coding_exon_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','5/12','',0,'SLC38A1','transcript',-1,''),
                        ('T','n.562G>A','ENST00000551506',-1,'non_coding_exon_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','5/5','',0,'SLC38A1','transcript',-1,''),
                        ('C','c.593C>G','ENST00000318737',-1,'stop_gained','protein_coding',false,'p.Ser198*','722','593','198',1444,4661,4335,'HIGH','6/26','',0,'C9orf84','transcript',-1,''),
                        ('C','c.476C>G','ENST00000394777',-1,'stop_gained','protein_coding',false,'p.Ser159*','721','476','159',1370,4955,4113,'HIGH','4/23','',0,'C9orf84','transcript',-1,''),
                        ('C','c.476C>G','ENST00000394779',-1,'stop_gained','protein_coding',false,'p.Ser159*','721','476','159',1405,5060,4218,'HIGH','4/24','',0,'C9orf84','transcript',-1,''),
                        ('C','c.593C>G','ENST00000374287',-1,'stop_gained','protein_coding',false,'p.Ser198*','854','593','198',1444,4847,4335,'HIGH','8/28','',0,'C9orf84','transcript',-1,''),
                        ('C','c.785C>G','ENST00000374283',-1,'stop_gained','protein_coding',false,'p.Ser262*','930','785','262',488,1973,1467,'HIGH','8/12','',0,'C9orf84','transcript',-1,''),
                        ('A','c.186G>C','ENST00000369454',-1,'synonymous_variant','protein_coding',false,'p.Ile62Ile','487','186','62',213,3505,642,'LOW','1/2','',0,'RAB39B','transcript',-1,''),
                        ('A','c.214G>C','ENST00000313654',-1,'missense_variant','protein_coding',false,'p.Gly72Ser','455','214','72',3333,10661,10002,'MODERATE','1/75','',0,'LAMA3','transcript',-1,''),
                        ('A','c.214G>A','ENST00000399516',-1,'missense_variant','protein_coding',false,'p.Gly72Ser','214','214','72',3277,9834,9834,'MODERATE','1/74','',0,'LAMA3','transcript',-1,''),
                        ('A','c.214G>A','ENST00000313654',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'LAMA3','domain:Laminin_N-terminal',-1,''),
                        ('A','c.214G>A','ENST00000399516',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'LAMA3','domain:Laminin_N-terminal',-1,''),
                        ('A','n.214G>A','ENST00000585600',-1,'non_coding_exon_variant','non_stop_decay',false,'','','','',-1,-1,-1,'MODIFIER','1/13','',0,'LAMA3','transcript',-1,''),
                        ('C','c.1445T>C','ENST00000233954',-1,'missense_variant','protein_coding',false,'p.Met482Thr','1716','1445','482',556,2058,1671,'MODERATE','11/11','',0,'IL1RL1','transcript',-1,''),
                        ('C','c.1445T>C','ENST00000233954',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'IL1RL1','topological_domain:Cytoplasmic',-1,''),
                        ('C','c.1445T>C','ENST00000233954',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'IL1RL1','domain:TIR',-1,''),
                        ('C','c.-10966T>C','ENST00000409599',-1,'upstream_gene_variant','protein_coding',false,'','','','',-1,-1,-1,'MODIFIER','','',0,'IL18R1','transcript',4234,''),
                        ('C','n.-4350T>C','ENST00000466357',-1,'upstream_gene_variant','retained_intron',false,'','','','',-1,-1,-1,'MODIFIER','','',0,'IL18R1','transcript',4350,''),
                        ('C','c.-28-10938T>C','ENST00000410040',-1,'intron_variant','protein_coding',false,'','','','',-1,-1,-1,'MODIFIER','1/10','',0,'IL18R1','transcript',-1,''),
                        ('A','c.1362G>A','ENST00000299440',-1,'synonymous_variant','protein_coding',false,'p.Leu454Leu','1474','1362','454',1043,6564,3132,'LOW','2/2','',0,'RAG1','transcript',-1,''),
                        ('A','c.1362G>A','ENST00000299440',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'RAG1','DNA-binding_region:NBD',-1,''),
                        ('A','n.*908C>T','ENST00000524423',-1,'downstream_gene_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','','',0,'RAG2','transcript',908,''),
                        ('A','n.1362G>A','ENST00000534663',-1,'non_coding_exon_variant','nonsense_mediated_decay',false,'','','','',-1,-1,-1,'MODIFIER','8/10','',0,'RAG1','transcript',-1,''),
                        ('T','c.1944G>T','ENST00000534015',-1,'splice_region_variant&synonymous_variant','protein_coding',false,'p.Pro648Pro','1945','1944','648',805,2787,2418,'LOW','15/20','',0,'NCAM1','transcript',-1,'WARNING_TRANSCRIPT_NO_START_CODON'),
                        ('T','c.1839G>T','ENST00000401611',-1,'splice_region_variant&synonymous_variant','protein_coding',false,'p.Thr613Thr','1839','1839','613',664,1995,1995,'LOW','16/18','',0,'NCAM1','transcript',-1,'WARNING_TRANSCRIPT_NO_START_CODON'),
                        ('T','n.447G>T','ENST00000526322',-1,'splice_region_variant&non_coding_exon_variant','processed_transcript',false,'','','','',-1,-1,-1,'LOW','6/10','',0,'NCAM1','transcript',-1,''),
                        ('T','n.342G>T','ENST00000531817',-1,'splice_region_variant&non_coding_exon_variant','nonsense_mediated_decay',false,'','','','',-1,-1,-1,'LOW','5/11','',0,'NCAM1','transcript',-1,''),
                        ('T','n.323G>T','ENST00000528590',-1,'splice_region_variant&non_coding_exon_variant','processed_transcript',false,'','','','',-1,-1,-1,'LOW','4/10','',0,'NCAM1','transcript',-1,''),
                        ('T','c.1973-5581G>T','ENST00000524665',-1,'intron_variant','protein_coding',false,'','','','',-1,-1,-1,'MODIFIER','13/18','',0,'NCAM1','transcript',-1,'WARNING_TRANSCRIPT_NO_START_CODON'),
                        ('T','n.2242-5581G>T','ENST00000531915',-1,'intron_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','15/20','',0,'NCAM1','transcript',-1,''),
                        ('T','n.2040+5623G>T','ENST00000527506',-1,'intron_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','13/15','',0,'NCAM1','transcript',-1,''),
                        ('T','c.1442-5581G>T','ENST00000533760',-1,'intron_variant','protein_coding',false,'','','','',-1,-1,-1,'MODIFIER','13/16','',0,'NCAM1','transcript',-1,''),
                        ('T','n.2037-5581G>T','ENST00000397957',-1,'intron_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','14/19','',0,'NCAM1','transcript',-1,''),
                        ('T','n.231-15090G>T','ENST00000526427',-1,'intron_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','1/3','',0,'NCAM1','transcript',-1,''),
                        ('T','n.2265-5581G>T','ENST00000531044',-1,'intron_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','13/19','',0,'NCAM1','transcript',-1,''),
                        ('T','c.1795+5623G>T','ENST00000316851',-1,'intron_variant','protein_coding',false,'','','','',-1,-1,-1,'MODIFIER','13/17','',0,'NCAM1','transcript',-1,''),
                        ('T','n.508+4472G>T','ENST00000530543',-1,'intron_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','5/6','',0,'NCAM1','transcript',-1,''),
                        ('T','n.323+5623G>T','ENST00000533073',-1,'intron_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','3/7','',0,'NCAM1','transcript',-1,''),
                        ('T','n.192+5623G>T','ENST00000525355',-1,'intron_variant','retained_intron',false,'','','','',-1,-1,-1,'MODIFIER','2/3','',0,'NCAM1','transcript',-1,''),
                        ('T','n.281G>T','ENST00000525691',-1,'non_coding_exon_variant','retained_intron',false,'','','','',-1,-1,-1,'MODIFIER','1/4','',0,'NCAM1','transcript',-1,''),
                        ('G','c.3226A>G','ENST00000506720',-1,'missense_variant','protein_coding',false,'p.Thr1076Ala','3226','3226','1076',1580,4962,4743,'MODERATE','17/25','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3022A>G','ENST00000512091',-1,'missense_variant','protein_coding',false,'p.Thr1008Ala','3769','3022','1008',1240,12636,3723,'MODERATE','18/26','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3022A>G','ENST00000514591',-1,'missense_variant','protein_coding',false,'p.Thr1008Ala','3351','3022','1008',1469,6297,4410,'MODERATE','18/25','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3022A>G','ENST00000506700',-1,'missense_variant','protein_coding',false,'p.Thr1008Ala','3195','3022','1008',1231,4838,3696,'MODERATE','16/23','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3226A>G','ENST00000509896',-1,'missense_variant','protein_coding',false,'p.Thr1076Ala','3399','3226','1076',1308,5069,3927,'MODERATE','17/25','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3226A>G','ENST00000511324',-1,'missense_variant','protein_coding',false,'p.Thr1076Ala','3399','3226','1076',1299,5042,3900,'MODERATE','17/24','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3022A>G','ENST00000545650',-1,'missense_variant','protein_coding',false,'p.Thr1008Ala','3195','3022','1008',1469,6124,4410,'MODERATE','16/23','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3226A>G','ENST00000507164',-1,'missense_variant','protein_coding',false,'p.Thr1076Ala','3355','3226','1076',1342,4982,4029,'MODERATE','17/25','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3226A>G','ENST00000508693',-1,'missense_variant','protein_coding',false,'p.Thr1076Ala','3355','3226','1076',1351,5009,4056,'MODERATE','17/26','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3226A>G','ENST00000507625',-1,'missense_variant','protein_coding',false,'p.Thr1076Ala','3332','3226','1076',1528,4843,4587,'MODERATE','17/23','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3022A>G','ENST00000504896',-1,'missense_variant','protein_coding',false,'p.Thr1008Ala','3117','3022','1008',1283,4816,3852,'MODERATE','16/25','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3022A>G','ENST00000514157',-1,'missense_variant','protein_coding',false,'p.Thr1008Ala','3117','3022','1008',1274,4789,3825,'MODERATE','16/24','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3226A>G','ENST00000506746',-1,'missense_variant','protein_coding',false,'p.Thr1076Ala','3226','3226','1076',1571,4935,4716,'MODERATE','17/24','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3022A>G','ENST00000508946',-1,'missense_variant','protein_coding',false,'p.Thr1008Ala','3022','3022','1008',1512,4758,4539,'MODERATE','16/24','',0,'LPHN3','transcript',-1,''),
                        ('G','c.3022A>G','ENST00000514996',-1,'missense_variant','protein_coding',false,'p.Thr1008Ala','3022','3022','1008',1503,4731,4512,'MODERATE','16/23','',0,'LPHN3','transcript',-1,''),
                        ('G','c.1393A>G','ENST00000502815',-1,'missense_variant','protein_coding',false,'p.Thr465Ala','1394','1393','465',917,4284,2754,'MODERATE','8/14','',0,'LPHN3','transcript',-1,'WARNING_TRANSCRIPT_NO_START_CODON'),
                        ('G','c.3022A>G','ENST00000512091',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'LPHN3','topological_domain:Extracellular',-1,''),
                        ('A','c.2219G>T','ENST00000272427',-1,'missense_variant','protein_coding',false,'p.Trp740Leu','2350','2219','740',811,5918,2436,'MODERATE','21/22','',0,'EXOC6B','transcript',-1,''),
                        ('A','n.158G>T','ENST00000490919',-1,'non_coding_exon_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','3/4','',0,'EXOC6B','transcript',-1,''),
                        ('A','n.140G>T','ENST00000471335',-1,'non_coding_exon_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','3/4','',0,'EXOC6B','transcript',-1,''),
                        ('A','n.183G>T','ENST00000492257',-1,'non_coding_exon_variant','processed_transcript',false,'','','','',-1,-1,-1,'MODIFIER','2/3','',0,'EXOC6B','transcript',-1,''),
                        ('G','c.1381C>G','ENST00000286398',-1,'missense_variant','protein_coding',false,'p.Leu461Val','1669','1381','461',1197,5976,3594,'MODERATE','11/25','',0,'SMC2','transcript',-1,''),
                        ('G','c.1381C>G','ENST00000374793',-1,'missense_variant','protein_coding',false,'p.Leu461Val','1594','1381','461',1197,5909,3594,'MODERATE','11/25','',0,'SMC2','transcript',-1,''),
                        ('G','c.1381C>G','ENST00000303219',-1,'missense_variant','protein_coding',false,'p.Leu461Val','1731','1381','461',1099,3650,3300,'MODERATE','11/24','',0,'SMC2','transcript',-1,''),
                        ('G','c.1381C>G','ENST00000374787',-1,'missense_variant','protein_coding',false,'p.Leu461Val','1717','1381','461',1197,4266,3594,'MODERATE','11/25','',0,'SMC2','transcript',-1,''),
                        ('G','c.1381C>G','ENST00000303219',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'SMC2','coiled-coil_region',-1,''),
                        ('G','c.1381C>G','ENST00000374787',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'SMC2','coiled-coil_region',-1,''),
                        ('C','c.586A>G','ENST00000306480',-1,'missense_variant','protein_coding',false,'p.Arg196Gly','732','586','196',271,6198,816,'MODERATE','5/6','',0,'TMEM192','transcript',-1,''),
                        ('C','c.574A>G','ENST00000506087',-1,'missense_variant','protein_coding',false,'p.Arg192Gly','757','574','192',267,2225,804,'MODERATE','6/7','',0,'TMEM192','transcript',-1,''),
                        ('C','c.163A>G','ENST00000505095',-1,'missense_variant','protein_coding',false,'p.Arg55Gly','698','163','55',83,789,254,'MODERATE','6/6','',0,'TMEM192','transcript',-1,'WARNING_TRANSCRIPT_INCOMPLETE'),
                        ('C','c.586A>G','ENST00000306480',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'TMEM192','topological_domain:Cytoplasmic',-1,''),
                        ('C','c.574A>G','ENST00000506087',-1,'sequence_feature','protein_coding',false,'','','','',-1,-1,-1,'LOW','','',0,'TMEM192','topological_domain:Cytoplasmic',-1,''),
                        ('T','c.3152G>A','ENST00000399788',-1,'missense_variant','protein_coding',false,'p.Arg1051Gln','3515','3152','1051',1690,10763,5073,'MODERATE','21/28','',0,'KDM5A','transcript',-1,''),
                        ('T','c.3152G>A','ENST00000382815',-1,'missense_variant','protein_coding',false,'p.Arg1051Gln','3515','3152','1051',1636,5274,4911,'MODERATE','21/28','',0,'KDM5A','transcript',-1,''),
                        ('T','c.2009G>A','ENST00000544760',-1,'missense_variant','protein_coding',false,'p.Arg670Gln','2200','2009','670',714,2338,2147,'MODERATE','12/13','',0,'KDM5A','transcript',-1,'WARNING_TRANSCRIPT_INCOMPLETE'),
                        ('T','n.-3176G>A','ENST00000540156',-1,'upstream_gene_variant','retained_intron',false,'','','','',-1,-1,-1,'MODIFIER','','',0,'KDM5A','transcript',3176,''),
                        ('T','n.266G>A','ENST00000535269',-1,'non_coding_exon_variant','retained_intron',false,'','','','',-1,-1,-1,'MODIFIER','1/2','',0,'KDM5A','transcript',-1,''),
                        ('A','c.1271G>A','ENST00000262794',-1,'missense_variant','protein_coding',false,'p.Arg424His','1354','1271','424',1211,3960,3636,'MODERATE','9/27','',0,'MOV10L1','transcript',-1,''),
                        ('A','c.1271G>A','ENST00000545383',-1,'missense_variant','protein_coding',false,'p.Arg424His','1355','1271','424',1211,3889,3636,'MODERATE','10/28','',0,'MOV10L1','transcript',-1,''),
                        ('A','c.1271G>A','ENST00000395858',-1,'missense_variant','protein_coding',false,'p.Arg424His','1295','1271','424',1165,3763,3498,'MODERATE','9/26','',0,'MOV10L1','transcript',-1,''),
                        ('A','c.1211G>A','ENST00000540615',-1,'missense_variant','protein_coding',false,'p.Arg404His','1431','1211','404',1165,3941,3498,'MODERATE','9/26','',0,'MOV10L1','transcript',-1,''),
                        ('A','c.-1436G>A','ENST00000395843',-1,'5_prime_UTR_variant','protein_coding',false,'','','','',-1,-1,-1,'MODIFIER','9/22','',0,'MOV10L1','transcript',32526,''),
                        ('A','n.308G>A','ENST00000434497',-1,'non_coding_exon_variant','nonsense_mediated_decay',false,'','','','',-1,-1,-1,'MODIFIER','3/7','',0,'MOV10L1','transcript',-1,'');

--
-- Name: ensembl; Type: TABLE; Schema: variantstore;
--

CREATE TABLE ensembl (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    version integer,
    date character varying,
    referencegenome_id bigint
);

INSERT INTO ensembl(version, date, referencegenome_id) VALUES (1, '01/01/2020', 1),
                                                              (2, '01/01/2021', 1);


--
-- Name: ensembl_gene; Type: TABLE; Schema: variantstore;
--

CREATE TABLE ensembl_gene (
    ensembl_id bigint NOT NULL,
    gene_id bigint NOT NULL
);

INSERT INTO ensembl_gene(ensembl_id, gene_id) VALUES (2, 1), (2,2),
                                                     (2,3),(2,4),
                                                     (2,5), (2,6),
                                                     (2,7), (2,8),
                                                     (2,9), (2,10),
                                                     (2,11), (2,12),
                                                     (2,13), (2,14),
                                                     (2,15), (2,16),
                                                     (2,17), (2,18),
                                                     (2,19), (2,20),
                                                     (2,21), (2,22),
                                                     (2,23), (2,24),
                                                     (2,25), (2,26),
                                                     (2,27), (2,28),
                                                     (2,29), (2,30),
                                                     (2,31), (2,32),
                                                     (2,33), (2,34),
                                                     (2,35), (2,36),
                                                     (2,37), (2,38),
                                                     (2,39), (2,40),
                                                     (2,41), (2,42),
                                                     (2,43), (2,44),
                                                     (2,45), (2,46),
                                                     (2,47), (2,48),
                                                     (2,49), (2,50),
                                                     (2,51), (2,52),
                                                     (2,53), (2,54),
                                                     (2,55), (2,56),
                                                     (2,57), (2,58),
                                                     (2,59), (2,60),
                                                     (2,61), (2,62),
                                                     (2,63), (2,64),
                                                     (2,65), (2,66),
                                                     (2,67), (2,68),
                                                     (2,69), (2,70),
                                                     (2,71), (2,72),
                                                     (2,73), (2,74),
                                                     (2,75), (2,76),
                                                     (2,77), (2,78),
                                                     (2,79), (2,80),
                                                     (2,81), (2,82),
                                                     (2,83), (2,84),
                                                     (2,85), (2,86),
                                                     (2,87), (2,88),
                                                     (2,89), (2,90),
                                                     (2,91);


--
-- Name: entity; Type: TABLE; Schema: variantstore;
-- id character varying NOT NULL,

CREATE TABLE entity (
    id bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    identifier character varying,
    project_id bigint NOT NULL
);

INSERT INTO entity(identifier, project_id) VALUES ('patient1', (select id FROM project WHERE identifier='PROJ')),
                                                  ('patient2', (select id FROM project WHERE identifier='PROJ')),
                                                  ('patient3', (select id FROM project WHERE identifier='PROJ'));

--
-- Name: gene; Type: TABLE; Schema: variantstore;
--

CREATE TABLE gene (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    symbol character varying,
    name character varying,
    biotype character varying,
    chr character varying,
    start bigint,
    "end" bigint,
    synonyms character varying[],
    geneid character varying,
    description character varying,
    strand character varying,
    version integer
);

INSERT INTO gene(symbol, name, biotype, chr, start, "end", synonyms, geneid, description, strand, version)
VALUES
       ('LAMA3',NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000053747',NULL,NULL,NULL),
       ('LAMA3',NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000053747',NULL,NULL,NULL),
       ('LAMA3',NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000053747',NULL,NULL,NULL),
       ('LAMA3',NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000053747',NULL,NULL,NULL),
       ('LAMA3',NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000053747',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073146',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073146',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073146',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073146',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073146',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073146',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073614',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073614',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073614',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073614',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000073614',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000111371',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000111371',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000111371',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000111371',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000111371',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000111371',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000111371',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000111371',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000115602',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000115602',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000115602',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000115604',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000115604',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000115604',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000136824',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000136824',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000136824',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000136824',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000136824',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000136824',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000144036',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000144036',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000144036',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000144036',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000149294',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000150471',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000155961',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000155961',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000165181',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000165181',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000165181',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000165181',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000165181',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000166349',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000166349',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000166349',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000170088',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000170088',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000170088',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000170088',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000170088',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000175097',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000248905',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000248905',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000248905',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000248905',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000248905',NULL,NULL,NULL),
       (NULL,NULL,NULL,NULL,NULL,NULL,NULL,'ENSG00000248905',NULL,NULL,NULL);

--
-- Name: gene_consequence; Type: TABLE; Schema: variantstore;
--

CREATE TABLE gene_consequence (
    consequence_id bigint NOT NULL,
    gene_id bigint NOT NULL
);

INSERT INTO gene_consequence(consequence_id, gene_id)
VALUES (1,1),(2,1),
       (3,1),(4,1),
       (5,1),(6,1),
       (7,7),(8,7),
       (9,7),(10,7),
       (11,7),(12,7),
       (13,7),(14,7),
       (15,15),(16,15),
       (17,15),(18,15),
       (19,15),(20,20),
       (21,21),(22,21),
       (23,21),(24,21),
       (25,21),(26,26),
       (27,26),(28,26),
       (29,29),(30,29),
       (31,29),(32,32),
       (33,32),(34,34),
       (35,32),(36,36),
       (37,36),(38,36),
       (39,36),(40,36),
       (41,36),(42,36),
       (43,36),(44,36),
       (45,36),(46,36),
       (47,36),(48,36),
       (49,36),(50,36),
       (51,36),(52,36),
       (53,53),(54,53),
       (55,53),(56,53),
       (57,53),(58,53),
       (59,53),(60,53),
       (61,53),(62,53),
       (63,53),(64,53),
       (65,53),(66,53),
       (67,53),(68,53),
       (69,53),(71,71),
       (72,71),(73,71),
       (74,71),(75,75),
       (76,75),(77,75),
       (78,75),(79,75),
       (80,75),(81,81),
       (82,81),(83,81),
       (84,81),(85,81),
       (86,86),(87,86),
       (88,86),(89,86),
       (90,86),(91,91),
       (92,91),(93,91),
       (94,91),(95,91),
       (96,91);

--
-- Name: genotype; Type: TABLE; Schema: variantstore;
--

CREATE TABLE genotype (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    genotype character varying,
    readdepth integer,
    filter character varying,
    likelihoods character varying,
    genotypelikelihoods character varying,
    genotypelikelihoodshet character varying,
    posteriorprobs character varying,
    genotypequality integer,
    haplotypequalities character varying,
    phaseset character varying,
    phasingquality integer,
    alternateallelecounts character varying,
    mappingquality integer
);

INSERT INTO genotype(genotype, readdepth, filter, likelihoods, genotypelikelihoods, genotypelikelihoodshet,
                     posteriorprobs, genotypequality, haplotypequalities, phaseset, phasingquality,
                     alternateallelecounts, mappingquality)
VALUES ('A', NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL),
       ('C', NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL),
       ('G', NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL);

--
-- Name: referencegenome; Type: TABLE; Schema: variantstore;
--

CREATE TABLE referencegenome (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    source character varying,
    build character varying,
    version character varying
);

INSERT INTO referencegenome(source, build, version) VALUES ('Ensembl','GRCh37','17');

--
-- Name: referencegenome_variant; Type: TABLE; Schema: variantstore;
--

CREATE TABLE referencegenome_variant (
    referencegenome_id bigint NOT NULL,
    variant_id bigint NOT NULL
);

--
-- Name: sample; Type: TABLE; Schema: variantstore;
--

CREATE TABLE sample (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    identifier character varying,
    cancerentity character varying,
    entity_id bigint
);

INSERT INTO sample(identifier, cancerentity, entity_id) VALUES ('QTEST001AL','HCC', (select id FROM entity WHERE identifier='patient1')),
                                                               ('QTEST002AT','BRCA', (select id FROM entity WHERE identifier='patient2')),
                                                               ('QTEST003A3','ALL', (select id FROM entity WHERE identifier='patient3'));

--
-- Name: sample_variant; Type: TABLE; Schema: variantstore;
--

CREATE TABLE sample_variant (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    sample_id bigint,
    variant_id bigint,
    vcfinfo_id bigint,
    genotype_id bigint
);

INSERT INTO sample_variant(sample_id, variant_id, vcfinfo_id, genotype_id)
VALUES (1,1,1,1),
       (1,2,2,1),
       (1,3,3,1),
       (1,4,4,1),
       (1,5,5,1),
       (2,4,4,1),
       (2,6,6,1),
       (2,7,7,1),
       (2,8,8,1),
       (2,9,9,1),
       (3,11,11,1),
       (3,12,12,1),
       (3,13,13,1),
       (3,14,14,1),
       (3,15,15,1);



--
-- Name: variant; Type: TABLE; Schema: variantstore;
--

CREATE TABLE variant (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    chr character varying,
    start bigint,
    "end" bigint,
    ref character varying,
    obs character varying,
    somatic boolean,
    identifier character varying,
    databaseidentifier character varying
);

INSERT INTO variant(chr, start, "end", ref, obs, somatic, identifier, databaseidentifier)
VALUES ('15',33359344,33359344,'T','C',true,'e741f1f2-2c39-4d00-b311-5bf4cf896005', '.'),
       ('12',46601390,46601390,'C','G',true,'d0b9a24d-e95a-44a2-8279-e895772f97cf','.'),
       ('9',114518682,114518682,'G','T',true,'aed9a27b-f9aa-4f8f-911f-81936aef306e','.'),
       ('X',154493388,154493388,'G','C',true,'c001f62d-5729-4dd8-9900-2f62d768a12d','.'),
       ('18',21269861,21269861,'G','C',true,'eb88a067-8c81-4c74-81e7-27d059c09add','.'),
       ('2',102968155,102968155,'T','C',true,'c68a421d-1028-405e-8ee0-17b6a250e1ad','.'),
       ('11',36596216,36596216,'G','A',true,'ccefd3cf-0737-4c73-bb1b-a3e45ab8d591','.'),
       ('11',113111509,113111509,'G','T',true,'ab938f60-1d0d-4124-a548-8a5b4575f1c8','.'),
       ('4',62849311,62849311,'A','G',true,'74632419-3371-4920-b9d6-638a7860fc64','.'),
       ('2',72411294,72411294,'C','A',true,'4ad08366-6b4c-4b98-8ea5-24a5637ca417','.'),
       ('9',106875723,106875723,'C','G',true,'9d9840fe-085d-4bc4-b9f1-a881662616d8','.'),
       ('4',166006829,166006829,'T','C',true,'7ddac11f-f383-462f-8c4a-4cefdbe79da5','.'),
       ('12',420115,420115,'C','T',true,'a44d09a9-198b-4849-a2ea-06cd004123a6','.'),
       ('22',50555597,50555597,'G','A',true,'ca32be21-bae0-46a2-9f7d-5a707256bc75','.');


INSERT INTO referencegenome_variant(referencegenome_id, variant_id)
VALUES ((select id FROM referencegenome WHERE build='GRCh37'),1),
       ((select id FROM referencegenome WHERE build='GRCh37'),2),
       ((select id FROM referencegenome WHERE build='GRCh37'),3),
       ((select id FROM referencegenome WHERE build='GRCh37'),4),
       ((select id FROM referencegenome WHERE build='GRCh37'),5),
       ((select id FROM referencegenome WHERE build='GRCh37'),6),
       ((select id FROM referencegenome WHERE build='GRCh37'),7),
       ((select id FROM referencegenome WHERE build='GRCh37'),8),
       ((select id FROM referencegenome WHERE build='GRCh37'),9),
       ((select id FROM referencegenome WHERE build='GRCh37'), 10),
       ((select id FROM referencegenome WHERE build='GRCh37'), 11),
       ((select id FROM referencegenome WHERE build='GRCh37'), 12),
       ((select id FROM referencegenome WHERE build='GRCh37'), 13),
       ((select id FROM referencegenome WHERE build='GRCh37'), 14),
       ((select id FROM referencegenome WHERE build='GRCh37'),15);
--
-- Name: variant_consequence; Type: TABLE; Schema: variantstore;
--

CREATE TABLE variant_consequence (
    variant_id bigint NOT NULL,
    consequence_id bigint NOT NULL
);

INSERT INTO variant_consequence(variant_id, consequence_id)
VALUES (1, 1),
       (1, 2),
       (1, 3),
       (1, 4),
       (1, 5),
       (1, 6),
       (2, 7),
       (2, 8),
       (2, 9),
       (2, 10),
       (2, 11),
       (2, 12),
       (2, 13),
       (2, 14),
       (3, 15),
       (3, 16),
       (3, 17),
       (3, 18),
       (3, 19),
       (4, 20),
       (5, 21),
       (5, 22),
       (5, 23),
       (5, 24),
       (5, 25),
       (6, 26),
       (6, 27),
       (6, 28),
       (6, 29),
       (6, 30),
       (6, 31),
       (7, 32),
       (7, 33),
       (7, 34),
       (7, 35),
       (8, 36),
       (8, 37),
       (8, 38),
       (8, 39),
       (8, 40),
       (8, 41),
       (8, 42),
       (8, 43),
       (8, 44),
       (8, 45),
       (8, 46),
       (8, 47),
       (8, 48),
       (8, 49),
       (8, 50),
       (8, 51),
       (8, 52),
       (9, 53),
       (9, 54),
       (9, 55),
       (9, 56),
       (9, 57),
       (9, 58),
       (9, 59),
       (9, 60),
       (9, 61),
       (9, 62),
       (9, 63),
       (9, 64),
       (9, 65),
       (9, 66),
       (9, 67),
       (9, 68),
       (9, 69),
       (11, 71),
       (11, 72),
       (11, 73),
       (11, 74),
       (12, 75),
       (12, 76),
       (12, 77),
       (12, 78),
       (12, 79),
       (12, 80),
       (13, 81),
       (13, 82),
       (13, 83),
       (13, 84),
       (13, 85),
       (14, 86),
       (14, 87),
       (14, 88),
       (14, 89),
       (14, 90),
       (15, 91),
       (15, 92),
       (15, 93),
       (15, 94),
       (15, 95),
       (15, 96);

--
-- Name: variantcaller; Type: TABLE; Schema: variantstore;
--

CREATE TABLE variantcaller (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    name character varying,
    version character varying,
    doi character varying
);

INSERT INTO variantcaller(name, version, doi) VALUES ('Strelka','2.0','10.1038/s41592-018-0051-x');

--
-- Name: variantcaller_variant; Type: TABLE; Schema: variantstore;
--

CREATE TABLE variantcaller_variant (
    variant_id bigint NOT NULL,
    variantcaller_id bigint NOT NULL
);

INSERT INTO variantcaller_variant(variant_id, variantcaller_id)
VALUES (1, 1),
       (2, 1),
       (3, 1),
       (4, 1),
       (5, 1),
       (6, 1),
       (7, 1),
       (8, 1),
       (9, 1),
       (10, 1),
       (11, 1),
       (12, 1),
       (13, 1),
       (14, 1);

--
-- Name: vcfinfo; Type: TABLE; Schema: variantstore;
--

CREATE TABLE vcfinfo (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 2147483647 CACHE 1 ),
    ancestralallele character varying,
    allelecount character varying,
    allelefrequency character varying,
    numberalleles integer,
    basequality integer,
    cigar character varying,
    dbsnp boolean,
    hapmaptwo boolean,
    hapmapthree boolean,
    thousandgenomes boolean,
    combineddepth integer,
    endpos integer,
    rms float8,
    mqzero integer,
    strandbias integer,
    numbersamples integer,
    somatic boolean,
    validated boolean
);

INSERT INTO vcfinfo(ancestralallele, allelecount, allelefrequency, numberalleles, basequality, cigar, dbsnp, hapmaptwo,
                    hapmapthree, thousandgenomes, combineddepth, endpos, rms, mqzero, strandbias, numbersamples,
                    somatic, validated)
VALUES ('p.E135K', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.G72S', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.I62I', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.L454L', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.L461V', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.M482T', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.N248H', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.R1051Q', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.R192G', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.R424H', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.S262*', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.T1008A', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.T613T', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false),
       ('p.W277L', '', '', -1, -1, '', false, false, false, false, -1, -1, -1.0, -1, -1, -1, false, false);

--
-- Name: annotationsoftware annotationsoftware_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY annotationsoftware
    ADD CONSTRAINT annotationsoftware_pkey PRIMARY KEY (id);

--
-- Name: consequence consequence_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY consequence
    ADD CONSTRAINT consequence_pkey PRIMARY KEY (id);


--
-- Name: ensembl ensembl_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY ensembl
    ADD CONSTRAINT ensembl_pkey PRIMARY KEY (id);


--
-- Name: entity entity_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY entity
    ADD CONSTRAINT entity_pkey PRIMARY KEY (id);


--
-- Name: gene gene_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY gene
    ADD CONSTRAINT gene_pkey PRIMARY KEY (id);


--
-- Name: genotype genotype_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY genotype
    ADD CONSTRAINT genotype_pkey PRIMARY KEY (id);


--
-- Name: project project_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY project
    ADD CONSTRAINT project_pkey PRIMARY KEY (id);


--
-- Name: referencegenome referencegenome_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY referencegenome
    ADD CONSTRAINT referencegenome_pkey PRIMARY KEY (id);


--
-- Name: sample sample_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY sample
    ADD CONSTRAINT sample_pkey PRIMARY KEY (id);


--
-- Name: variant variant_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY variant
    ADD CONSTRAINT variant_pkey PRIMARY KEY (id);


--
-- Name: variantcaller variantcaller_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY variantcaller
    ADD CONSTRAINT variantcaller_pkey PRIMARY KEY (id);


--
-- Name: vcfinfo vcfinfo_pkey; Type: CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY vcfinfo
    ADD CONSTRAINT vcfinfo_pkey PRIMARY KEY (id);


--
-- Name: as_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX as_unique ON annotationsoftware USING btree (name, version, doi);


--
-- Name: ascons_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX ascons_unique ON annotationsoftware_consequence USING btree (annotationsoftware_id, consequence_id);


--
-- Name: cons_gene_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX cons_gene_unique ON gene_consequence USING btree (gene_id, consequence_id);


--
-- Name: cons_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX cons_unique ON consequence USING btree (allele, codingchange, transcriptid, transcriptversion, type, biotype, canonical, aachange, cdnaposition, cdsposition, proteinposition, proteinlength, cdnalength, cdslength, impact, exon, intron, strand, genesymbol, featuretype, distance);


--
-- Name: ensembl_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX ensembl_unique ON ensembl USING btree (version, date, referencegenome_id);

--
-- Name: gene_unique; Type: INDEX; Schema: variantstore;
--

CREATE INDEX gene_unique ON gene USING btree (geneid);

--
-- Name: genotype_unique; Type: INDEX; Schema: variantstore;
--

CREATE INDEX genotype_unique ON genotype USING btree (genotype, readdepth, filter, likelihoods, genotypelikelihoods, genotypelikelihoodshet, posteriorprobs, genotypequality, haplotypequalities, phaseset, phasingquality, alternateallelecounts, mappingquality);

--
-- Name: idx_chromosome; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_chromosome ON variant USING btree (chr);


--
-- Name: idx_constype; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_constype ON consequence USING btree (type);


--
-- Name: idx_ensembl_id; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_ensembl_id ON ensembl_gene USING btree (ensembl_id);


--
-- Name: idx_entity_id; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_entity_id ON sample USING btree (entity_id);


--
-- Name: idx_gene_id; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_gene_id ON ensembl_gene USING btree (gene_id);


--
-- Name: idx_geneid; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_geneid ON gene USING btree (geneid);


--
-- Name: idx_genesymbol; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_genesymbol ON consequence USING btree (genesymbol);


--
-- Name: idx_rg; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_rg ON ensembl USING btree (referencegenome_id);


--
-- Name: idx_rg_build; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_rg_build ON referencegenome USING btree (build);


--
-- Name: idx_sample_id; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_sample_id ON sample USING btree (identifier);


--
-- Name: idx_sample_variant; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_sample_variant ON sample_variant USING btree (sample_id, variant_id);


--
-- Name: idx_start; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_start ON variant USING btree (start);


--
-- Name: idx_variant_sample; Type: INDEX; Schema: variantstore;
--

CREATE INDEX idx_variant_sample ON sample_variant USING btree (variant_id, sample_id);


--
-- Name: rg_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX rg_unique ON referencegenome USING btree (source, build, version);


--
-- Name: sample_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX sample_unique ON sample USING btree (identifier, cancerentity, entity_id);

--
-- Name: project_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX project_unique ON project USING btree (identifier);

--
-- Name: entity_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX entity_unique ON entity USING btree (identifier, project_id);

--
-- Name: sample_variant_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX sample_variant_unique ON sample_variant USING btree (sample_id, variant_id, vcfinfo_id, genotype_id);


--
-- Name: variant_rg_unique; Type: INDEX; Schema: variantstore;
--

-- CREATE UNIQUE INDEX variant_rg_unique ON referencegenome_variant USING btree (referencegenome_id, variant_id);


--
-- Name: variant_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX variant_unique ON variant USING btree (chr, start, "end", ref, obs, somatic, databaseidentifier);


--
-- Name: variant_vc_unique; Type: INDEX; Schema: variantstore;
--

-- CREATE UNIQUE INDEX variant_vc_unique ON variantcaller_variant USING btree (variant_id, variantcaller_id);


--
-- Name: variantcaller_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX variantcaller_unique ON variantcaller USING btree (name, version, doi);


--
-- Name: vcfinfo_unique; Type: INDEX; Schema: variantstore;
--

CREATE UNIQUE INDEX vcfinfo_unique ON vcfinfo USING btree (ancestralallele, allelecount, allelefrequency, numberalleles, basequality, cigar, dbsnp, hapmaptwo, hapmapthree, thousandgenomes, combineddepth, endpos, rms, mqzero, strandbias, numbersamples, somatic, validated);

-- CREATE UNIQUE INDEX variant_consequence_unique ON variant_consequence USING btree (variant_id, consequence_id);


--
-- Name: annotationsoftware_consequence annotationsoftware_consequence_annotationsoftware_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY annotationsoftware_consequence
    ADD CONSTRAINT annotationsoftware_consequence_annotationsoftware_id_fkey FOREIGN KEY (annotationsoftware_id) REFERENCES annotationsoftware(id) NOT VALID;


--
-- Name: annotationsoftware_consequence annotationsoftware_consequence_consequence_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY annotationsoftware_consequence
    ADD CONSTRAINT annotationsoftware_consequence_consequence_id_fkey FOREIGN KEY (consequence_id) REFERENCES consequence(id) NOT VALID;


--
-- Name: ensembl_gene ensembl_gene_ensembl_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY ensembl_gene
    ADD CONSTRAINT ensembl_gene_ensembl_id_fkey FOREIGN KEY (ensembl_id) REFERENCES ensembl(id) NOT VALID;


--
-- Name: ensembl_gene ensembl_gene_gene_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY ensembl_gene
    ADD CONSTRAINT ensembl_gene_gene_id_fkey FOREIGN KEY (gene_id) REFERENCES gene(id) NOT VALID;


--
-- Name: ensembl ensembl_referencegenome_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY ensembl
    ADD CONSTRAINT ensembl_referencegenome_id_fkey FOREIGN KEY (referencegenome_id) REFERENCES referencegenome(id) NOT VALID;


--
-- Name: entity entity_project_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY entity
    ADD CONSTRAINT entity_project_id_fkey FOREIGN KEY (project_id) REFERENCES project(id) NOT VALID;


--
-- Name: gene_consequence gene_consequence_consequence_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY gene_consequence
    ADD CONSTRAINT gene_consequence_consequence_id_fkey FOREIGN KEY (consequence_id) REFERENCES consequence(id) NOT VALID;


--
-- Name: gene_consequence gene_consequence_gene_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY gene_consequence
    ADD CONSTRAINT gene_consequence_gene_id_fkey FOREIGN KEY (gene_id) REFERENCES gene(id) NOT VALID;


--
-- Name: referencegenome_variant referencegenome_variant_referencegenome_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY referencegenome_variant
    ADD CONSTRAINT referencegenome_variant_referencegenome_id_fkey FOREIGN KEY (referencegenome_id) REFERENCES referencegenome(id) NOT VALID;


--
-- Name: referencegenome_variant referencegenome_variant_variant_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY referencegenome_variant
    ADD CONSTRAINT referencegenome_variant_variant_id_fkey FOREIGN KEY (variant_id) REFERENCES variant(id) NOT VALID;


--
-- Name: sample sample_entity_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY sample
    ADD CONSTRAINT sample_entity_id_fkey FOREIGN KEY (entity_id) REFERENCES entity(id) NOT VALID;


--
-- Name: sample_variant sample_variant_genotype_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY sample_variant
    ADD CONSTRAINT sample_variant_genotype_id_fkey FOREIGN KEY (genotype_id) REFERENCES genotype(id) NOT VALID;


--
-- Name: sample_variant sample_variant_sample_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY sample_variant
    ADD CONSTRAINT sample_variant_sample_id_fkey FOREIGN KEY (sample_id) REFERENCES sample(id) NOT VALID;


--
-- Name: sample_variant sample_variant_variant_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY sample_variant
    ADD CONSTRAINT sample_variant_variant_id_fkey FOREIGN KEY (variant_id) REFERENCES variant(id) NOT VALID;


--
-- Name: sample_variant sample_variant_vcfinfo_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY sample_variant
    ADD CONSTRAINT sample_variant_vcfinfo_id_fkey FOREIGN KEY (vcfinfo_id) REFERENCES vcfinfo(id) NOT VALID;


--
-- Name: variantcaller_variant variantcaller_variant_variant_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY variantcaller_variant
    ADD CONSTRAINT variantcaller_variant_variant_id_fkey FOREIGN KEY (variant_id) REFERENCES variant(id) NOT VALID;

--
-- Name: variantcaller_variant variantcaller_variant_variantcaller_id_fkey; Type: FK CONSTRAINT; Schema: variantstore;
--

ALTER TABLE ONLY variantcaller_variant
    ADD CONSTRAINT variantcaller_variant_variantcaller_id_fkey FOREIGN KEY (variantcaller_id) REFERENCES variantcaller(id) NOT VALID;

ALTER TABLE ONLY variant_consequence
    ADD CONSTRAINT variant_consequence_variant_id_fkey FOREIGN KEY (variant_id) REFERENCES variant(id) NOT VALID;

ALTER TABLE ONLY variant_consequence
    ADD CONSTRAINT variant_consequence_consequence_id_fkey FOREIGN KEY (consequence_id) REFERENCES consequence(id) NOT VALID;


--
-- PostgreSQL database dump complete
--
